lib/Sema/TypeCheckStmt.cpp

```diff
@@ -442,6 +442,30 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
       return RS;
     }
 
+    // If the body consisted of a single return without a result
+    // 
+    //   func foo() -> Int {
+    //     return
+    //   }
+    // 
+    // in parseAbstractFunctionBody the return is given an empty, implicit tuple
+    // as its result
+    //
+    //   func foo() -> Int {
+    //     return ()
+    //   }
+    //
+    // Look for that case and diagnose it as missing return expression.
+    if (!ResultTy->isVoid() && TheFunc->hasSingleExpressionBody()) {
+      auto expr = TheFunc->getSingleExpressionBody();
+      if (expr->isImplicit() &&
+          dyn_cast<TupleExpr>(expr) != nullptr &&
+          dyn_cast<TupleExpr>(expr)->getNumElements() == 0) {
```

Comment:
This isn't what I meant, but I think it works. You can simplify to `isa<TupleExpr>(expr) && cast<TupleExpr>(expr)->`, though.

Reply:
That is much nicer!  Fixed in https://github.com/apple/swift/pull/25802/commits/bdf1c0b243978103b58b026c3a90edfbd934f80f .

Reply:
`cast` was correct, by the way. It's an asserting version of `dyn_cast`, but since you just checked with `isa` it's fine. (I guess it'll be optimized either way, but it's still convention.)

Reply:
Oops.  I meant to commit that change (as the commit message shows) but failed to actually commit it.  Fixed in https://github.com/apple/swift/pull/25802/commits/6dab25bb3e39010aea76a8f6f5b236a8f83f0a5f .  And thank you for the explanation!